app/queries/new_accounts_query.rb

```diff
@@ -72,4 +52,36 @@ def granularity_to_date_format(granularity)
       raise "Unknown granularity #{granularity.inspect}"
     end
   end
+
+  def mysql_subquery(range, date_format, timezone: Time.zone.tzinfo.name)
+    account.buyer_accounts
+        .where.has { sift(:date, sift(:in_timezone, created_at, name: timezone)).in(range) }
+        .grouping { sift(:date_format, sift(:in_timezone, created_at, name: timezone), date_format).to_sql }
+        .count(:id)
+  end
+
+  def oracle_subquery(range, date_format, timezone: Time.zone.tzinfo.name)
+    query = account.buyer_accounts.where.has do
+      sift(:date, sift(:in_timezone, created_at, name: timezone)).in(range)
+    end
+
+    query = query.selecting do
+      [
+        id,
+        sift(:date_format, sift(:in_timezone, created_at, name: timezone), date_format).as('dategrouping')
+      ]
+    end
+
+    Account.from(query, 'subquery').group('dategrouping').count('subquery.id')
+  end
+
+  def query_with_offset_retry(range, date_format)
+    time_zone = Time.zone
+
+    begin
+      yield range, date_format, timezone: time_zone.tzinfo.name
+    rescue ActiveRecord::StatementInvalid
```

Comment:
Rescuing from `ActiveRecord::StatementInvalid` is not recommended with Postgres. See https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html (section _Exception handling and rolling back_)

Reply:
Fixed for postgres. Added a comment to address later for Oracle.